// react-map-gl/dist/es5/components/scale-control.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf3 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _propTypes = _interopRequireDefault(require("prop-types"));

var _baseControl = _interopRequireDefault(require("./base-control"));

var _mapboxgl = _interopRequireDefault(require("../utils/mapboxgl"));

var propTypes = Object.assign({}, _baseControl["default"].propTypes, {
  maxWidth: _propTypes["default"].number,
  unit: _propTypes["default"].oneOf(['imperial', 'metric', 'nautical'])
});
var defaultProps = Object.assign({}, _baseControl["default"].defaultProps, {
  maxWidth: 100,
  unit: 'metric'
});

var ScaleControl = function (_BaseControl) {
  (0, _inherits2["default"])(ScaleControl, _BaseControl);

  function ScaleControl() {
    var _getPrototypeOf2;

    var _this;

    (0, _classCallCheck2["default"])(this, ScaleControl);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = (0, _possibleConstructorReturn2["default"])(this, (_getPrototypeOf2 = (0, _getPrototypeOf3["default"])(ScaleControl)).call.apply(_getPrototypeOf2, [this].concat(args)));
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_control", null);
    (0, _defineProperty2["default"])((0, _assertThisInitialized2["default"])(_this), "_mapboxScaleControl", null);
    return _this;
  }

  (0, _createClass2["default"])(ScaleControl, [{
    key: "componentDidMount",
    value: function componentDidMount() {
      var mapboxScaleControl = new _mapboxgl["default"].ScaleControl();
      mapboxScaleControl._map = this._context.map;
      mapboxScaleControl._container = this._containerRef.current;
      this._mapboxScaleControl = mapboxScaleControl;

      this._update();
    }
  }, {
    key: "_update",
    value: function _update() {
      var mapboxScaleControl = this._mapboxScaleControl;

      if (mapboxScaleControl) {
        mapboxScaleControl.options = this.props;

        mapboxScaleControl._onMove();
      }
    }
  }, {
    key: "_render",
    value: function _render() {
      this._control = this._control || _react["default"].createElement("div", {
        ref: this._containerRef,
        className: "mapboxgl-ctrl mapboxgl-ctrl-scale"
      });

      this._update();

      return this._control;
    }
  }]);
  return ScaleControl;
}(_baseControl["default"]);

exports["default"] = ScaleControl;
(0, _defineProperty2["default"])(ScaleControl, "propTypes", propTypes);
(0, _defineProperty2["default"])(ScaleControl, "defaultProps", defaultProps);
