// @vtex/react-csv-parse/es/index.js
import _default5 from 'ramda/src/splitAt';
import _default4 from 'ramda/src/zipObj';
import _default3 from 'ramda/src/lift';
import _default2 from 'ramda/src/apply';
import _default from 'ramda/src/compose';

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

import React from 'react';
import PropTypes from 'prop-types';

import Papa from 'papaparse';

var CsvParse = function (_React$Component) {
  _inherits(CsvParse, _React$Component);

  function CsvParse() {
    var _temp, _this, _ret;

    _classCallCheck(this, CsvParse);

    for (var _len = arguments.length, args = Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    return _ret = (_temp = (_this = _possibleConstructorReturn(this, _React$Component.call.apply(_React$Component, [this].concat(args))), _this), _this.handleFile = function (event) {
      var file = event.target.files[0];
      var keys = _this.props.keys;
      var onDataUploaded = _this.props.onDataUploaded;
      var onError = _this.props.onError;

      Papa.parse(file, {
        skipEmptyLines: true,
        error: function error(err, file, inputElem, reason) {
          onError({ err: err, file: file, inputElem: inputElem, reason: reason });
        },
        complete: function complete(results) {
          var data = results.data;

          // remove display headers
          data.shift();

          // add api headers
          data.unshift(keys);

          // convert arrays to objects
          var formatedResult = _default(_default2(_default3(_default4)), _default5(1))(data);

          // send result to state
          onDataUploaded(formatedResult);
        }
      });
    }, _temp), _possibleConstructorReturn(_this, _ret);
  }

  CsvParse.prototype.render = function render() {
    return this.props.render(this.handleFile);
  };

  return CsvParse;
}(React.Component);

CsvParse.propTypes = process.env.NODE_ENV !== "production" ? {
  keys: PropTypes.array.isRequired,
  onDataUploaded: PropTypes.func.isRequired,
  onError: PropTypes.func
} : {};

export default CsvParse;
